add a readme file for the thumbnails service (#4737)
owncloud/ocis: services/thumbnails/README.md

# Thumbnails service

> The thumbnails service provides methods to retrieve thumbnails for various files.
> It loads the source files from the oCIS storage and stores the generated thumbnails locally on disk.

Reviewer: ```suggestion
The thumbnails service provides methods to generate thumbnails for various files.
It retrieves the sources at the location where the user files are stored and saves the tumbnails at the location where system files are stored. Those locations have defaults but can be manually defined via environment variables. The relevant environment variables are `OCIS_BASE_DATA_PATH` which will contain system relevant data and `STORAGE_USERS_OCIS_ROOT` which will contain the user source data. For details and defaults see the documentation.
```

Author: > `THUMBNAILS_FILESYSTEMSTORAGE_ROOT` used if thumnails should be separated from system files.

Isn't this missing word(s) before used?

Thumbnails can be generated for the following file types:

  - png
  - jpg
  - gif
  - tiff
  - bmp
  - txt

> And the thumbnails can either be generated as `png` or `jpg` files.

Author: The generated image types can be requested by the caller via a query parameter.

> And the thumbnails can either be generated as `png` or `jpg` files.

Reviewer: Which means that always both image types are generated which contradicts the generate part of the sentence above - or?

> And the thumbnails can either be generated as `png` or `jpg` files.

Reviewer: How is that done that a thumb can be requested as a specific type?

# Thumbnails service

The thumbnails service provides methods to generate thumbnails for various files.
It retrieves the sources at the location where the user files are stored and saves the tumbnails at the location where system files are stored. Those locations have defaults but can be manually defined via environment variables. The relevant environment variables are:

-   `OCIS_BASE_DATA_PATH` which will contain system relevant data and
-   `STORAGE_USERS_OCIS_ROOT` which will contain the user source data.
-   `THUMBNAILS_FILESYSTEMSTORAGE_ROOT` used if thumnails should be separated from system files.

For details and defaults see the documentation.

> Thumbnails can be generated for the following file types:

Reviewer: Q1: are we identifying the image type by extension or by magic bytes or both?
Q2: If the image type is identified by extension, what happens if the file is not of the type?

Author: > Are we identifying the source image type by extension or by magic bytes or both?

The thumbnail service is just using the information given by the backend. The backend is identifying it based on the extension.

> If the source image type is identified by extension, what happens if the file is not of the type? Means extension ≠ magic bytes?

Thumbnail generation will fail.

-  png
-  jpg
-  gif
-  tiff
-  bmp
-  txt

> Thumbnails can either be generated as `png`, `jpg` or `gif` files, various resolutions can be defined via `THUMBNAILS_RESOLUTIONS`.

Reviewer: > The generated image types can be requested by the caller via a parameter.

* Which means that always both image types are generated which contradicts the generate part of the sentence above - or?
* How is that done that a thumb can be requested as a specific type?

Author: See my answer here: https://github.com/owncloud/ocis/pull/4737#discussion_r987044139

Reviewer: ok, I read it, I have to refine the question,

> The caller can specify the thumbnail type

1. Who is the "caller" and how does he know about the image types available?
2. The available image types are hardcoded in the thumbnail service, no env availabe - or?
3. The request can contain a resolution (W x H), we have an env defining the available resolutions. How does the caller know the available resolutions to select one available for the request?

`message GetThumbnailRequest` has a well defined content for the thumbnail service to generate the requested thumbnail. In a nutshell, how does the caller know about the options to define? and how does the caller identify the set to request?

Author: > Who is the "caller" and how does he know about the thumb types available?

It's another service or later it could be a client like the Desktop client or an App.
The available types are defined in the protobuf service definition

> Just to be sure, the available thumb types are hardcoded in the thumbnail service as described in the example abobe, no env availabe - or?

Yes, they are defined in the service definition and currently not configurable.

> The request can contain a resolution (W x H), we have an env defining the available resolutions. How does the caller know the available resolutions to select one available for the request?

The caller doesn't. It can request any arbitrary resolution and the thumbnail service will use the one closest to the requested resolution.
E.g.:

Requested: 18x12
Available: 30x20, 15x10, 9x6
Returned: 15x10

Reviewer: I will add some text about the resolutions.
But I am still not clear about what the circumstances are, that a service choses one type over another - out of the available. I got it that the service knows because of the definition, it now can select one out of the available, but how is the criteria defined to select one?

Author: > I got it that the service knows the available because of the definition, it now can select one, but what is the criteria to select one?

That is up to the developer of the calling service. The service could always request `png` files or it could always request `jpg` files because of smaller file sizes or it can make it dependent on some other criteria.

# Thumbnails Service

> The thumbnails service provides methods to generate thumbnails for various files and resolutions based on requests. It retrieves the sources at the location where the user files are stored and saves the tumbnails where system files are stored. Those locations have defaults but can be manually defined via environment variables.

Reviewer 2: ```suggestion
The thumbnails service provides methods to generate thumbnails for various files and resolutions based on requests. It retrieves the sources at the location where the user files are stored and saves the thumbnails where system files are stored. Those locations have defaults but can be manually defined via environment variables.
```

## File Locations Overview

The relevant environment variables defining file locations are:

-   (1) `OCIS_BASE_DATA_PATH`
-   (2) `STORAGE_USERS_OCIS_ROOT`
-   (3) `THUMBNAILS_FILESYSTEMSTORAGE_ROOT`

> (1) ... Having a default set by the ocis code, but if defined, used as base path for other services.  

Reviewer 3: ```suggestion
(1) ... Having a default set by the Infinite Scale code, but if defined, used as base path for other services.  
```

(2) ... Source files, defaults to (1) plus path component, but can be freely defined if required.  
(3) ... Target files, defaults to (1) plus path component, but can be freely defined if required.  

For details and defaults for these environment variables see the ocis admin documentation.

## Thumbnail Location

> It may be benificial to define the location of the thumbnails other than at system files. This is due the fact that storing thumbnails can consume a lot of space over time which not necessarily needs to reside on the same partition or mount or expensive drives.

Reviewer 2: ```suggestion
It may be beneficial to define the location of the thumbnails to be other than the default (with system files). This is due the fact that storing thumbnails can consume a lot of space over time which not necessarily needs to reside on the same partition or mount or expensive drives.
```

## Thumbnail Source File Types

Thumbnails can be generated from the following source file types:

-   png
-   jpg
-   gif
-   tiff
-   bmp
-   txt

The thumbnail service retrieves source files using the information provided by the backend. The Linux backend identifies source files usually based on the extension.

If a file type was not properly assigned or the type identification failed, thumbnail generation will fail and an error will be logged.

## Thumbnail Target File Types

> Thumbnails can either be generated as `png`, `jpg` or `gif` files. These types are hardcoded and no other types can be requested. A requestor like another service or a client can request one of the avialable types to be generated. If more than one type is required, each type must be requested individaully.

Reviewer 2: ```suggestion
Thumbnails can either be generated as `png`, `jpg` or `gif` files. These types are hardcoded and no other types can be requested. A requestor, like another service or a client, can request one of the available types to be generated. If more than one type is required, each type must be requested individually.
```

## Thumbnail Resolution

> Various resolutions can be defined via `THUMBNAILS_RESOLUTIONS`. A requestor can request any arbitrary resolution and the thumbnail service will use the one closest to the requested resolution. If more than one resolution is required, each resolution must be requested individaully.

Reviewer 2: ```suggestion
Various resolutions can be defined via `THUMBNAILS_RESOLUTIONS`. A requestor can request any arbitrary resolution and the thumbnail service will use the one closest to the requested resolution. If more than one resolution is required, each resolution must be requested individually.
```

> **Example:**

Reviewer: ```suggestion
Example:
```

Requested: 18x12  
Available: 30x20, 15x10, 9x6  
Returned: 15x10  

## Deleting Thumbnails

> As of now, there is no automatism deleting thumbnails automatically. This is especially true when a source file gets deleted or moved. This situation will be solved in a later stage. For the time being, if you run short on physical thumbnails space, you have to manually delete the thumbnail store to free space. Thumbnails will then be recreated on request.

Reviewer 2: ```suggestion
As of now, there is no automated thumbnail deletion. This is especially true when a source file gets deleted or moved. This situation will be solved at a later stage. For the time being, if you run short on physical thumbnails space, you have to manually delete the thumbnail store to free space. Thumbnails will then be recreated on request.
```

I guess that means that there is also no DELETE thumbnail API method - so clients (or nerds using `curl`) also cannot decide to delete thumbnails?

## Memory Considerations

> Since source files need to be loaded into memory when generating thumbnails, large source files could potentially crash this service if there is insufficient memory available. For bigger instances when using container orchestration deployment methods, this service can be dedicated to own servers with more memory.

Reviewer 2: ```suggestion
Since source files need to be loaded into memory when generating thumbnails, large source files could potentially crash this service if there is insufficient memory available. For bigger instances when using container orchestration deployment methods, this service can be dedicated to its own server(s) with more memory.
```
